src/pairwise.jl

```diff
@@ -46,76 +74,59 @@ function check_vectors(x, y, skipmissing::Symbol)
     end
 end
 
-function _pairwise!(::Val{:pairwise}, f, dest::AbstractMatrix, x, y, symmetric::Bool)
-    check_vectors(x, y, :pairwise)
-    @inbounds for (j, yj) in enumerate(y)
-        ynminds = .!ismissing.(yj)
-        @inbounds for (i, xi) in enumerate(x)
-            symmetric && i > j && continue
+function _pairwise!(::Val{:pairwise}, f, dest::AbstractMatrix{V}, x, y, symmetric::Bool) where {V}
 
-            if xi === yj
-                ynm = view(yj, ynminds)
-                # For performance, diagonal is special-cased
-                if f === cor && eltype(dest) !== Union{} && i == j
-                    # TODO: float() will not be needed after JuliaLang/Statistics.jl#61
-                    dest[i, j] = float(cor(xi))
+    nr, nc = size(dest)
+    m = length(x) == 0 ? 0 : length(first(x))
+
+    # Swap x and y for more efficient threaded loop.
+    if nr < nc
+        dest′ = reshape(dest, size(dest, 2), size(dest, 1))
+        _pairwise!(Val(:pairwise), f, dest′, y, x, symmetric)
+        dest .= transpose(dest′)
+        return dest
+    end
+
+    #cor and friends are special-cased.
+    iscor = (f in (corkendall, corspearman, cor))
+    (iscor || f == cov) && (symmetric = x === y)
+    #cov(x) is faster than cov(x, x)
+    (f == cov) && (f = ((x, y) -> x === y ? cov(x) : cov(x, y)))
+
+    nmtx = promoted_nmtype(x)[]
+    nmty = promoted_nmtype(y)[]
+
+    Threads.@threads for subset in equal_sum_subsets(nr, Threads.nthreads())
+        for j in subset
+            scratch_fx = task_local_vector(:scratch_fx, nmtx, m)
+            scratch_fy = task_local_vector(:scratch_fy, nmty, m)
+            for i = 1:(symmetric ? j : nc)
+                if iscor && i == j && y[i] === x[j] && V !== Union{} && V!== Missing
+                    dest[j, i] = 1.0
```

Comment:
Also here, I think it would be more correct to use something like
```suggestion
                    dest[j, i] = oneunit(f_mn)
```
(which would also allow to remove the checks for `V !== Union{}` and `V !== Missing`)